PODAUTO-89: Update repo to operator-sdk v1.35

maxcao13 opened the pull request:
This PR migrates the repo to a modern operator-sdk v1.35 setup from the previous hacked together setup. This was done by first following the steps in this migration guide that John linked: https://sdk.operatorframework.io/docs/building-operators/golang/migration/, and then porting over the old repo's tests and other configuration for backwards compat.

Here's a summary of the list of changes:
* Operator-sdk migration-specific
  * Migrated types from `pkg/apis` to `api`.
  * Migrated reconcile code to the new manager format
  * Updated layout of the repo to fit the operator-sdk layout (config with Kustomize, bundle builds, etc.)
  * Updated Dockerfiles
  * Migrated old build and deploy steps
    * There is now a simple `deploy` make target, `deploy-bundle` target, where you can deploy the operator with OLM, and a `deploy-catalog` target where you can deploy a `CatalogSource` + `OperatorGroup` with the operator contained to see the UI steps that someone may take to create the operator in the OCP console OperatorHub  
  * full list can be found in the link above
* VPA-specific
  * Removed `hack/manifest-diff.sh` and its jq helper since we no longer need to sync install manifests with our CSV (operator-sdk does it for us!)
  * Updated `update-vendor.sh` to fit the new repo
  * Removed `hack/go-*.sh` since operator-sdk generated Makefile has these commands already
  * Updated `hack/e2e.sh` to include a flag where you can use your own temporary directory to pull vpa tests from (this was because my internet was being bad and I found it annoying to keep cloning 80 MB)
  * Created `config/olm-catalog` directory to contain the yamls that are required for the catalog build
  * Created `config/vpa` directory with various configs which is required for the vpa operand to function

This will definitely require some additional changes in Prow config and probably in here as well to make that work.
 
Please let me know if I missed anything :-)

maxcao13 commented:
Needs some additional fixes that I haven't pushed yet, but e2e tests look good!
![image](https://github.com/user-attachments/assets/0a1120b9-773a-479a-b29e-3bf1c5b78e85)

maxcao13 commented:
Another commit is just going to be a README update, which can detail some of the testing steps anyone can take to review.

maxcao13 commented:
My `gitignore` was seriously ignoring a vendored directory and I didn't notice... :/

maxcao13 commented:
Reherasals have passed here: https://github.com/openshift/release/pull/55425

maxcao13 commented on internal/operator/status.go:
```diff
@@ -74,9 +76,7 @@ func (r *StatusReporter) SetRelatedObjects(objs []configv1.ObjectReference) {
-	for _, obj := range objs {
-		r.config.RelatedObjects = append(r.config.RelatedObjects, obj)
-	}
+	r.config.RelatedObjects = append(r.config.RelatedObjects, objs...)
```
A lot of these changes in the code was because of some `staticcheck` warnings from golangci-lint. There was one warning I haven't fixed yet because I'm not sure how, so I am ignoring it for now, and I hope we can just skip the `lint` test in the CI.

maxcao13 replied:
Thanks for the tip! Most(?) of the lint changes here I manually changed after reading the error, I didn't do `lint:fix` despite in case it would change something crazy :stuck_out_tongue:

maxcao13 commented on Makefile:
```diff
@@ -1,323 +1,442 @@
+	mkdir -p $(OLM_OUTPUT_DIR)
+	cp -r $(OLM_MANIFESTS_DIR)/* $(OLM_OUTPUT_DIR)
+
+	cd $(OLM_OUTPUT_DIR) && $(KUSTOMIZE) edit set image quay.io/openshift/origin-vertical-pod-autoscaler-operator-catalog=$(CATALOG_IMG)
```
I didn't want to edit all these kustomize files when deploying since it would cause a lot of git noise if you are just testing it, so I copy the directory to an output folder like we've done before, and edit everything there.

maxcao13 reviewed (commented):
I'm leaving a lot of comments to justify some of the changes and hopefully aid the review.

maxcao13 commented on Dockerfile:
```diff
@@ -1,12 +1,34 @@
+# was called. For example, if we call make docker-build in a local env which has the Apple Silicon M1 SO
+# the docker BUILDPLATFORM arg will be linux/arm64 when for Apple x86 it will be linux/amd64. Therefore,
+# by leaving it empty we can ensure that the container and binary shipped on it will have the same platform.
+RUN CGO_ENABLED=0 GOOS=${TARGETOS:-linux} GOARCH=${TARGETARCH} go build -mod=vendor -a -o manager cmd/main.go
```
I tried with a build without `CGO_ENABLED`, and it seemed to work as expected, so I can change this if that's what we want.

jkyros commented on Makefile:
```diff
@@ -1,323 +1,442 @@
+.PHONY: predeploy
+predeploy: ## Setup configuration for the operator before deployment.
+	cd config/manager && $(KUSTOMIZE) edit set image quay.io/openshift/origin-vertical-pod-autoscaler-operator=$(OPERATOR_IMG)
+	sed -i '$(VPA_REPLACE_LINE)s|value:.*|value: $(OPERAND_IMG)|' $(VPA_OPERAND_REPLACE_FILE)
```
Hahaha so wait:
- we ultimately need to overide the operand value so we make it an ENV variable 
- but in order for that to happen, we need kustomize to rewrite `manager.yaml` with that ENV variable
- but in order for that to happen, we need to add a patch to the `kustomization.yaml` patch list containing that ENV variable
- but the patch will be wrong unless we replace the value of that ENV variable in the kustomization patch using sed ?

Would this be less fragile if it the sed replacement was anchored on matching the value directly instead of the line number? Or is there a reason we can't do that? 

Or....maybe is there a way we could do it with `kustomize edit remove patch` / `kustomize edit add patch`  or something?   I understand how hacky/unwieldy kustomize is, theory vs practice, etc so the answer might be "yes but no"  :stuck_out_tongue: )

maxcao13 replied:
Yes, unfortunately that's what is happening :smiling_face_with_tear:. I don't remember why I didn't just sed directly and skip the 10 other steps, but I think I tunneled on trying to solve this with Kustomize. It's probably doable with a patch file, without sed now that i think about it. I'll try it out.

maxcao13 replied:
Okay, I think I was able to make it cleaner with `kustomize edit remove/add patch` and `yq`.

jkyros replied:
Makes sense! Thanks!  

I think it's totally fine how you're downloading `yq` right now, but it is a go program, so you could probably skip all that if you wanted to and do like a `go install github.com/mikefarah/yq/v4@latest` ( this is not "plz change it" this is "fyi" :smile: )

jkyros commented on hack/yaml-lint.sh:
```diff
@@ -1,11 +0,0 @@
-#!/bin/sh
```
It didn't look like we actually ran it anywhere? Like if it actually worked/ran/passed that seems nice to have so we know if we goofed, but I didn't see CI call it, and it doesn't seem like it was wired up to anything. Did I miss it?

maxcao13 replied:
No I don't think it was running in CI either :confused:. What do you think we should do? I'm with just skipping it.

jkyros replied:
hahaha well the last time we fixed anything it caught was 3 years ago according to blame, and we're currently naughty in the main branch. 

And since we're copying those vpa manifests from `kuberenetes-autoscaler`, messing with the yaml here is just going to ruin the diff. But that still leaves everything we don't copy. 

How about...we leave it out and make a JIRA card to tweak the rules and re-add yaml-lint.sh and a makefile target someday? 
```
install/deploy/01_vpacontroller.crd.yaml
  27:121    warning  line too long (171 > 120 characters)  (line-length)
  32:121    warning  line too long (175 > 120 characters)  (line-length)
  91:121    warning  line too long (156 > 120 characters)  (line-length)
  94:121    warning  line too long (154 > 120 characters)  (line-length)
  101:121   warning  line too long (156 > 120 characters)  (line-length)
  161:121   warning  line too long (156 > 120 characters)  (line-length)
  164:121   warning  line too long (154 > 120 characters)  (line-length)
  171:121   warning  line too long (156 > 120 characters)  (line-length)
  231:121   warning  line too long (156 > 120 characters)  (line-length)
  234:121   warning  line too long (154 > 120 characters)  (line-length)
  241:121   warning  line too long (156 > 120 characters)  (line-length)

install/deploy/05_vpa-crd.yaml
  121:121   warning  line too long (150 > 120 characters)  (line-length)
  132:121   warning  line too long (150 > 120 characters)  (line-length)
  292:121   warning  line too long (150 > 120 characters)  (line-length)
  305:121   warning  line too long (150 > 120 characters)  (line-length)
  314:121   warning  line too long (150 > 120 characters)  (line-length)
  330:121   warning  line too long (150 > 120 characters)  (line-length)
  401:121   warning  line too long (150 > 120 characters)  (line-length)
  412:121   warning  line too long (150 > 120 characters)  (line-length)
  532:121   warning  line too long (150 > 120 characters)  (line-length)
  545:121   warning  line too long (150 > 120 characters)  (line-length)
  554:121   warning  line too long (150 > 120 characters)  (line-length)
  570:121   warning  line too long (150 > 120 characters)  (line-length)

manifests/stable/vertical-pod-autoscaler-controller.crd.yaml
  27:121    warning  line too long (171 > 120 characters)  (line-length)
  32:121    warning  line too long (175 > 120 characters)  (line-length)
  91:121    warning  line too long (156 > 120 characters)  (line-length)
  94:121    warning  line too long (154 > 120 characters)  (line-length)
  101:121   warning  line too long (156 > 120 characters)  (line-length)
  161:121   warning  line too long (156 > 120 characters)  (line-length)
  164:121   warning  line too long (154 > 120 characters)  (line-length)
  171:121   warning  line too long (156 > 120 characters)  (line-length)
  231:121   warning  line too long (156 > 120 characters)  (line-length)
  234:121   warning  line too long (154 > 120 characters)  (line-length)
  241:121   warning  line too long (156 > 120 characters)  (line-length)

manifests/stable/vpa-v1.crd.yaml
  121:121   warning  line too long (150 > 120 characters)  (line-length)
  132:121   warning  line too long (150 > 120 characters)  (line-length)
  292:121   warning  line too long (150 > 120 characters)  (line-length)
  305:121   warning  line too long (150 > 120 characters)  (line-length)
  314:121   warning  line too long (150 > 120 characters)  (line-length)
  330:121   warning  line too long (150 > 120 characters)  (line-length)
  401:121   warning  line too long (150 > 120 characters)  (line-length)
  412:121   warning  line too long (150 > 120 characters)  (line-length)
  532:121   warning  line too long (150 > 120 characters)  (line-length)
  545:121   warning  line too long (150 > 120 characters)  (line-length)
  554:121   warning  line too long (150 > 120 characters)  (line-length)
  570:121   warning  line too long (150 > 120 characters)  (line-length)
```

maxcao13 replied:
Sure sounds good :) I can put in a ticket later.

jkyros reviewed (commented):
This looks really good! I don't really see anything that's not going to work, I just had a few questions around integrations and why some of the things changed.

maxcao13 commented on api/v1/verticalpodautoscalercontroller_types.go:
```diff
@@ -78,29 +74,56 @@ type ContainerOverride struct {
-	// TODO: Add status fields.
+// VerticalPodAutoscalerControllerSpec defines the desired state of VerticalPodAutoscalerController
+type VerticalPodAutoscalerControllerSpec struct {
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Safety Margin Fraction",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:number"}
```
Yeah, I'm pretty sure that Go names for these structs doesn't really matter. Only the structs that have an associated Kind in our API matters probably if we change the name. Yeah and we only refer to it as Spec anyways, so I think it's safe to change so we are consistent :)

maxcao13 commented on config/manager/manager.yaml:
```diff
@@ -7,24 +19,26 @@ metadata:
   labels:
     k8s-app: vertical-pod-autoscaler-operator
 spec:
-  replicas: 1
```
I thought about that, but I think we were not being deployed by CVO, but I wasn't completely sure. I was reading this: https://docs.ci.openshift.org/docs/how-tos/onboarding-a-new-component/#product-builds-and-becoming-part-of-an-openshift-release and it says that operators with this label are considered second-level operators (which I assume means shipped by CVO). I think we previously had this commented :laughing: https://github.com/openshift/vertical-pod-autoscaler-operator/blob/e498971fd4166261ffd5890e8d50aa118b123128/Dockerfile.rhel7#L12

maxcao13 commented on config/olm-catalog/vpa-subscription.yaml:
```diff
@@ -1,12 +1,10 @@
-  namespace: openshift-vertical-pod-autoscaler
+  name: vertical-pod-autoscaler-operator
 spec:
-  channel: OPERATOR_PACKAGE_CHANNEL
```
Apparently, previously we were replacing it here: https://github.com/openshift/vertical-pod-autoscaler-operator/blob/e498971fd4166261ffd5890e8d50aa118b123128/Makefile#L245

It doesn't really matter I think. These files are just for local deployment :)

jkyros replied:
Ok cool, just making sure. Thanks! :smile:

maxcao13 commented on hack/update-vendor.sh:
```diff
@@ -3,9 +3,11 @@
+# See https://sdk.operatorframework.io/docs/overview/#kubernetes-version-compatibility
+kube_release="v0.30.3"
 # these components k8s.io/<item> are versioned for each k8s release
-kube_components="api apiextensions-apiserver apimachinery client-go component-base"
```
Yep. Exactly.

maxcao13 commented on internal/operator/config.go:
```diff
@@ -93,7 +93,7 @@ func ConfigFromEnvironment() *Config {
 		config.VerticalPodAutoscalerName = caName
 	}
 
-	if caImage, ok := os.LookupEnv("VERTICAL_POD_AUTOSCALER_IMAGE"); ok {
```
Yep! It's here: https://github.com/openshift/vertical-pod-autoscaler-operator/pull/169/files#diff-ce0d597a9e5f20a8fe61c3ba5e6185a6b90cfef2c00bb5b5ef3e97b15b79f480R53

maxcao13 commented on config/scorecard/patches/olm.config.yaml:
```diff
@@ -0,0 +1,51 @@
+    labels:
+      suite: olm
+      test: olm-crds-have-validation-test
+# - TODO(macao): This fails since VPA checkpoint does not have resources in CSV. Figure out how to fix this.
```
No, I don't think it recommended anything. But AFAIK the checkpoint is simply just a store of historical values? https://povilasv.me/vertical-pod-autoscaling-the-definitive-guide/. I don't really know if that corresponds to concrete k8s resources it could own, so that's why I'm not sure if we should just skip this test.

jkyros replied:
Yeah, having run it, that test is way less great than I had hoped. I think those fields are ultimately advisory? It looks like even if you feed it "pods" or something for the checkpionts, it's also grumbly that the `v1beta2` version of the `VerticalPodAutoscaler` doesn't have any resources either. 

I don't think it matters. I think leaving it commented out for now is perfect.

maxcao13 commented on test/helpers/helpers.go:
```diff
@@ -2,7 +2,7 @@ package helpers
 import (
 	configv1 "github.com/openshift/api/config/v1"
-	"github.com/openshift/vertical-pod-autoscaler-operator/pkg/util"
+	"github.com/openshift/vertical-pod-autoscaler-operator/internal/util"
```
For some reason I thought it was, but going back to several different versions migration docs, I guess not. But either way, I agree no one should probably be depending on our code, so sticking it in `internal` should be okay to do.

maxcao13 replied:
I was digging through operator-sdk docs and noticed a lot of docs that referenced `controllers` in the root directory and `main.go` in the root directory as well, so I thought I was going crazy and was about to try migrating again. But I realized after doing an `operator-sdk init` on an empty directory that it scaffolds the `cmd/main.go` for us, and `operator-sdk create api` also generates `internal/controller` and sticks them in there.

So I think it's just a docs issue. I opened an issue on their upstream to confirm this.

maxcao13 commented on .gitignore:
```diff
@@ -1,19 +1,110 @@
-# ignore the binaries and goland configurations
+# Kubernetes Generated files - skip generated files, except for vendored files
+
+!vendor/**/zz_generated.*
```
I think I just copied this gitignore and forgot about it. I can remove the last part of that comment, and that whole `!vendor...` line altogether.

<sub>[I copied from my old team](https://github.com/cryostatio/cryostat-operator/blob/main/.gitignore) :) </sub>

maxcao13 commented:
Not really sure why `golint` still exists since I replaced it with `lint`. I'm gonna assume it's stale since we changed CI in the middle of this PR.

Also, `lint` is expected to fail because of that one deprecated function I mentioned earlier. :-)

jkyros commented:
> Also, lint is expected to fail because of that one deprecated function I mentioned earlier. :-) [Link to failed test](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_vertical-pod-autoscaler-operator/169/pull-ci-openshift-vertical-pod-autoscaler-operator-master-lint/1828602531387805696#1:build-log.txt%3A35) 

If we're squeamish about possible consequences from something like e.g. https://github.com/jkyros/vertical-pod-autoscaler-operator/commit/849ad7eebd82ef187f3056dcff0aa276d4af172e to fix the lint deprecation, can we maybe at least tag that line with a `TODO` comment and a `//nolint: staticcheck` so we don't have to skip the test in the mean time while we fix it?

jkyros commented:
Tests passed - awesome! Thanks for all your hard work here Max! 
/lgtm
